mantidproject/mantid
# New IDAaaS data cache index file  loading

### Description of work
Implemented search in data cache for files before defaulting to search in the data archive. This implementation leaves the previous implementation mostly untouched, and only adds a search in the data cache before advancing to the search in the archive. So the order for searching is User Directories > Data Cache > Data archive. This gives priority to local changes that users have made.

#### Summary of work
Added `ISISInstrDataCache.h` and `ISISInstrDataCache.cpp` which is where the implementation resides. Other than that, changed `FileFinder.cpp` to call the search. Added unit tests.

#### Purpose of work
To give access to data files through the data cache on IDAaaS even when the user does not have access to the archive.

Fixes #37013 

#### Further detail of work

Added a few comments and print statements to make it clear what is going on. We can remove them during the review stage.

### To test:
Build to test: https://builds.mantidproject.org/job/build_packages_from_branch/792/

1. Log into IDAaaS
2. Create a new conda environment and install build with `mamba install -c mantid/label/idaas_data_cache_debug mantidworkbench`
3. Open workbench
4. Open the file `/data/instrument/ALF/ALF_index.json`. The numbers on the left side correspond to available run numbers
5. In the workbench, change the default instrument to ALF and turn off access to the data archive.
6. Open the Load GUI and type in several run numbers of your choice chosen from the `ALF_index.json` file. You don't need to click `Run`, if the `OutputWorkspace` field gets updated then the file was successfully found. Try this several times and check they all successfully found the file.
7. Try some bad inputs (the error message will be displayed when hovering over the red asterisk):
    - A run number that doesn't exist, for example `99999`
    - Filename with wrong format, eg. `ALF`, `1234ALF`
    - File we don't have access to, eg. `MAR27340`







---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
